File: 슬라이딩 윈도우/leetcode438(Medium)-Find All Anagrams in a String.py
# leetcode 567 문제와 동일한 문제임
# 딕셔너리 해시테이블 풀이 : 333ms(40.9%), 15.2MB(74.41%)
class Solution:
    def lower_alphabets(self):
        return [chr(i) for i in range(97, 123)]

    def is_matched(self, a_map, b_map):
        for key in self.lower_alphabets():
            if a_map[key] != b_map[key]:
                return False
        return True

    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(p) > len(s):
            return []
            
        s_map = collections.defaultdict(int)
        p_map = collections.defaultdict(int)

        for i in range(len(p)):
            s_map[s[i]] += 1
            p_map[p[i]] += 1

        result = []
        for i in range(len(s) - len(p)):
            if self.is_matched(s_map, p_map):
                result.append(i)
            s_map[s[i]] -= 1
            s_map[s[i + len(p)]] += 1

        if self.is_matched(s_map, p_map):
            result.append(len(s) - len(p))

        return result

File: 슬라이딩 윈도우/leetcode438(Medium)-Find All Anagrams in a String.js
// leetcode 567 문제와 동일한 문제임
// 딕셔너리 해시테이블 풀이 : 363ms(26.78%), 48.5MB(53.93%)
const lowerAlphabets = () => {
  return Array.from({ length: 26 }, (v, i) => String.fromCharCode(i + 97))
}

const createAlphaMap = () => {
  return lowerAlphabets().reduce((obj, alpha) => {
    obj[alpha] = 0
    return obj
  }, {})
}

const isMatched = (aMap, bMap) => {
  for (const key of lowerAlphabets()) {
    if (aMap[key] !== bMap[key]) {
      return false
    }
  }
  return true
}

/**
 * @param {string} s
 * @param {string} p
 * @return {number[]}
 */
var findAnagrams = function (s, p) {
  if (p.length > s.length) {
    return []
  }

  const sMap = createAlphaMap()
  const pMap = createAlphaMap()

  for (let i = 0; i < p.length; i += 1) {
    sMap[s[i]] += 1
    pMap[p[i]] += 1
  }

  const result = []
  for (let i = 0; i < s.length - p.length; i += 1) {
    if (isMatched(sMap, pMap)) {
      result.push(i)
    }
    sMap[s[i]] -= 1
    sMap[s[i + p.length]] += 1
  }

  if (isMatched(sMap, pMap)) {
    result.push(s.length - p.length)
  }

  return result
}
